core/lib/vm_utils/src/vm_env.rs

```diff
@@ -0,0 +1,185 @@
+use multivm::{
+    interface::{L1BatchEnv, L2BlockEnv, SystemEnv, TxExecutionMode},
+    vm_latest::constants::BLOCK_GAS_LIMIT,
+};
+use zksync_contracts::BaseSystemContracts;
+use zksync_dal::StorageProcessor;
+use zksync_types::{
+    fee_model::BatchFeeInput, Address, L1BatchNumber, L2ChainId, MiniblockNumber,
+    ProtocolVersionId, H256, ZKPORTER_IS_AVAILABLE,
+};
+
+/// The environment required to initialize the VM.
+#[derive(Clone, Debug)]
+pub struct VmEnv {
+    pub l1_batch_env: L1BatchEnv,
+    pub system_env: SystemEnv,
+}
+
+/// The builder for the `VmEnv`. It allows to load the required data from the database.
+#[derive(Clone, Debug)]
+pub struct VmEnvBuilder {
+    miniblock_number: Option<MiniblockNumber>,
+    l1_batch_number: L1BatchNumber,
+    prev_batch_hash: Option<H256>,
+    fee_account: Option<Address>,
+    validation_computational_gas_limit: u32,
+    chain_id: L2ChainId,
+}
+
+impl VmEnvBuilder {
+    pub fn new(
+        l1_batch_number: L1BatchNumber,
+        validation_computational_gas_limit: u32,
+        chain_id: L2ChainId,
+    ) -> Self {
+        Self {
+            l1_batch_number,
+            validation_computational_gas_limit,
+            chain_id,
+            miniblock_number: None,
+            prev_batch_hash: None,
+            fee_account: None,
+        }
+    }
+    pub fn with_miniblock_number(mut self, miniblock_number: MiniblockNumber) -> Self {
+        self.miniblock_number = Some(miniblock_number);
+        self
+    }
+
+    pub fn with_fee_account(mut self, fee_account: Address) -> Self {
+        self.fee_account = Some(fee_account);
+        self
+    }
+
+    pub fn with_prev_batch_hash(mut self, prev_batch_hash: H256) -> Self {
+        self.prev_batch_hash = Some(prev_batch_hash);
+        self
+    }
+
+    pub async fn build(self, connection: &mut StorageProcessor<'_>) -> Option<VmEnv> {
```

Comment:
Nit: It could make sense to propagate errors / add contexts for errors here instead of `unwrap()`ping them.

Reply:
Yeah, probably and we could also return some kind of error if we found some discrepancy